<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [2]:
import pandas as pd
import psycopg2

In [3]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME='project_sql'
USER='skillfactory'
HOST='84.201.134.129'
PASSWORD='cCkxxLVrDE8EbvjueeMedPKt'
PORT='5432'

In [120]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [5]:
# текст запроса
query_3_1 = f'''select 
                    count(v.id) "Количество вакансий"
                from public.vacancies v 
            
            '''

In [6]:
# результат запроса
vac_es = pd.read_sql_query(query_3_1, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\3056453078.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_3_1, connection)


Количество вакансий
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [7]:
# текст запроса
query_3_2 = f'''select
                    count(e.id) "Количество работодателей"
                    from public.employers e 
            
            '''

In [8]:
# результат запроса
vac_es = pd.read_sql_query(query_3_2, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\1976313957.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_3_2, connection)


Количество работодателей
0                     23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [9]:
# текст запроса
query_3_3 = f'''select
                count(areas.id) "Количество регионов"
                from public.areas 
            
            '''

In [10]:
# результат запроса
vac_es = pd.read_sql_query(query_3_3, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\2712792629.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_3_3, connection)


Количество регионов
0                 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [11]:
# текст запроса
query_3_4 = f'''select
                    count(i.id) "Кол-во сфер деятельности"
                    from public.industries i 
            
            '''

In [12]:
# результат запроса
vac_es = pd.read_sql_query(query_3_4, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\1307782639.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_3_4, connection)


Кол-во сфер деятельности
0                       294

***

In [13]:
# выводы по предварительному анализу данных

# выводы по предварительному анализу данных
В выборке представлено 23501 работодателей, которые разместили 49197 вакансий в 1362 регионах. Вакансии представлены в дастаточно широком спектре сфер деятельности - 294

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [14]:
# текст запроса
n = 5
query_4_1 = f'''select 
                    a.name area,
                    count(v.id) cnt   
               from  
                    public.vacancies v 
                    full join public.areas a on v.area_id = a.id   --соединяем таблицу vacancies и areas
               group by 1                                          --групируем по названию региона сортируем по убыванию
               order by 2 desc
               limit {n}                                           --выводим первые 5 строк
            '''

In [15]:
# результат запроса
vac_es = pd.read_sql_query(query_4_1, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\1256599242.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [16]:
# текст запроса
query_4_2 = f'''select 
                    count(*) "Кол-во вакансий"                              --количество вакансий с хотя бы одним заполненным полем с зарплатой                     
                from 
                    public.vacancies v
                where v.salary_from is not null or salary_to is not null    -- условие по хотя бы одному заполненному  полю

             '''  
               
           

In [17]:
# результат запроса
vac_es = pd.read_sql_query(query_4_2, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\2546982240.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_4_2, connection)


Кол-во вакансий
0            24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [18]:
# текст запроса
query_4_3 = f'''select 
                    round(avg(salary_from)) average_low,    -- средняя зарплата для нижней границы
                    round(avg(salary_to)) average_upper     -- средняя зарплата для верхней границы
                from 
                    public.vacancies 
                     
               
               
               
         '''

In [19]:
# результат запроса
vac_es = pd.read_sql_query(query_4_3, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\182738571.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_4_3, connection)


average_low  average_upper
0      71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [20]:
# текст запроса
query_4_4 = f'''select 
                    schedule "Рабочий график",
                    employment "Тип трудоустройства",
                    count(id) "К-во вакансий"
                from 
                    public.vacancies 
                group by   1, 2                     -- группируем
                order by 3 desc                     -- сортируем по убыванию
                offset 1                            -- отсекаем 1 строку
                limit 1                             -- выводим первой стороку (второе место)
               
            '''

In [21]:
# результат запроса
vac_es = pd.read_sql_query(query_4_4, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\2749995665.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_4_4, connection)


Рабочий график Тип трудоустройства  К-во вакансий
0  Удаленная работа    Полная занятость           7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [22]:
# текст запроса
query_4_5 = f'''select 
                    experience "Опыт работы",
                    count(id) 
                from 
                    public.vacancies
                group by 1              -- группируем
                order by 2              -- сортировка по возрастанию к-ва вакансий
                     
              '''

In [23]:
# результат запроса
vac_es = pd.read_sql_query(query_4_5, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\2632085953.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_4_5, connection)


Опыт работы  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

In [24]:
# выводы по детальному анализу вакансий

# выводы по детальному анализу вакансий
Лидерами среди регионов по количеству размещенных вакансий являются мегаполисы с развитой промышленностью и хозяйственной инфраструктурой для которых требуется большое количество персонала. Почти у половины вакансий не указан размер зарплаты. Среднее значение по нижней границе ЗП - 71065 по верхней границе ЗП - 110537. Наиболее востребованные работники - работники на полный рабочий день с поной занятностью с опытом работы от 1 до 3 лет

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [25]:
# текст запроса

query_5_1 = f'''(select 
                    e.name "работодатель",
                    count(v.id) "к-во вакансий"    
                from 
                    public.vacancies v
                join public.employers e on v.employer_id = e.id
                group by 1              -- группируем
                order by 2 desc         -- сортируем в порядке убывания по кол-ву вакансий
                limit 1)                -- выводим работодателя на 1 месте
                
                union all               -- объединяем таблицы чтобы вывести работодателей на 1 и 5 месте
                
                (select 
                    e.name, 
                    count(v.id)   
                from 
                    public.vacancies v
                join public.employers e on v.employer_id = e.id
                group by 1
                order by 2 desc 
                offset 4                -- отсекаем первые 4 строки
                limit 1)
               '''

In [26]:
# результат запроса
vac_es = pd.read_sql_query(query_5_1, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\2549523760.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_5_1, connection)


работодатель  к-во вакансий
0         Яндекс           1933
1  Газпром нефть            331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [27]:
# текст запроса
query_5_2 = f'''select 
                    a.name "Регион",
                    count(e.id) "Кол-во работодателей"                   
                     
                from 
                    public.areas a
                    full join public.employers e on a.id = e.area
                    full join public.vacancies v on a.id = v.area_id
                group by 1 
                having count(v.id) = 0              -- выделяем регионы где нет вакансий 
                order by 2 desc                     -- сортируем по убыванию кол-ва работодателей  
                limit 1                             -- оставляем максимальное значение            
            '''         
               

In [28]:
# результат запроса
vac_es = pd.read_sql_query(query_5_2, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\726689865.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_5_2, connection)


Регион  Кол-во работодателей
0  Россия                   410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [29]:
# текст запроса
query_5_3 = f'''select
                    e.name "Работодатель",
                    count(distinct v.area_id) "Кол-во регионов"                 --считаем только уникальные регионы
                from 
                    public.vacancies v
                    full join public.employers e on v.employer_id = e.id 
                group by 1  
                order by 2 desc
                limit 1                                                         -- оставляем максимальное значение
            '''   

In [30]:
# результат запроса
vac_es = pd.read_sql_query(query_5_3, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\2482784884.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_5_3, connection)


Работодатель  Кол-во регионов
0       Яндекс              181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [31]:
# текст запроса
query_5_4 = f'''select
                    count(*) "Кол-во работодателей"
                from 
                    public.employers e
                    left join public.employers_industries ei on e.id = ei.employer_id
                where industry_id is null           -- условие отсутствия сферы деятельности

            '''


In [32]:
# результат запроса
vac_es = pd.read_sql_query(query_5_4, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\730612324.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_5_4, connection)


Кол-во работодателей
0                  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [33]:
# текст запроса
query_5_5 = f'''select
                    e.name "Название работодателя",
                    count(*) "Кол-во работодателей"
                from 
                    public.employers e
                    left join public.employers_industries ei on e.id = ei.employer_id
                group by 1
                having  count(*) = 4            -- условие на наличие 4 сфер деятельности
                order by 1
                offset 2                        -- отрезаем первые 2 строки
                limit 1                         -- выводим 1 (3) строку

            '''

In [34]:
# результат запроса
vac_es = pd.read_sql_query(query_5_5, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\3828865776.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_5_5, connection)


Название работодателя  Кол-во работодателей
0                  2ГИС                     4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [35]:
# текст запроса
query_5_5 = f'''select
                    count(ei.employer_id) "Кол-во работодателей"
                from 
                    public.employers_industries ei
                    join public.industries i on ei.industry_id = i.id
                where i.name = 'Разработка программного обеспечения'
                

            '''

In [36]:
# результат запроса
vac_es = pd.read_sql_query(query_5_5, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\3828865776.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_5_5, connection)


Кол-во работодателей
0                  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [37]:
# код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
tables = pd.read_html(url)
city_df = tables[1]
city_list = list(city_df['Город'])
display(city_list)

['Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград']

In [102]:
# текст запроса
query_5_7 = f'''select
                    a.name "Регионы-миллионники",           --регионы в которы представлена компания "Яндекс"
                    count(v.id) "Кол-во вакансий"           --кол-во вакансий "Яндекса" в этих регионах
                from 
                    public.vacancies v
                    left join public.employers e  on v.employer_id = e.id
                    left join public.areas a on v.area_id = a.id
                where e.name = 'Яндекс' and a.name in {tuple(city_list)}
                group by a.name
                
                union all
                
                select
                    'Total',                                -- добавляем строку с общим количеством вакансий
                    count(*)                    
                from 
                    public.vacancies v
                    left join public.employers e  on v.employer_id = e.id
                    left join public.areas a on v.area_id = a.id
                where e.name = 'Яндекс' and a.name in {tuple(city_list)}   
                order by 2 
             '''   

In [103]:
# результат запроса
vac_es = pd.read_sql_query(query_5_7, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\3196144504.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_5_7, connection)


Регионы-миллионники  Кол-во вакансий
0                 Омск               21
1            Челябинск               22
2           Красноярск               23
3            Волгоград               24
4                Пермь               25
5               Казань               25
6       Ростов-на-Дону               25
7                  Уфа               26
8               Самара               26
9            Краснодар               30
10             Воронеж               32
11         Новосибирск               35
12     Нижний Новгород               36
13        Екатеринбург               39
14     Санкт-Петербург               42
15              Москва               54
16               Total              485

***

In [40]:
# выводы по анализу работодателей

# выводы по анализу работодателей
По количеству вакансий лидирует компания "Яндекс" 1933 вакансии примерно 4% от общего количества. Среди регионов попадаются те в которых нет вакансий, но есть работодатели. По количеству регионов в которых размещены вакансии компании, ожидаемо лидирует "Яндекс", на втором месте компания "Ростелеком". Возможно это связано с тем, что сферы деятельности "Яндекса" в основном связаны с IT и это позволяет иметь достаточно большой штат сотрудников на "удаленке" в различных регионах и даже странах, а у "Ростетелекома" развитая сеть представительств (которые физически существуют) по всей России. Количество компаний у которых указано 4 сферы деятельности 1136. Компаний, которые занимаются разработкой программного обеспечения, 3553.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [62]:
# текст запроса
query_6_1 = f'''select
                    count(v.name) "Вакансии, связанные с данными"
                from public.vacancies v
                where lower(v.name) like '%data%' or lower(v.name) like '%данн%'
            '''

In [65]:
# результат запроса
vac_es = pd.read_sql_query(query_6_1, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\3254502547.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_6_1, connection)


Вакансии, связанные с данными
0                           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [66]:
# текст запроса
query_6_2 = f'''select
                    count(v.id) as "Вакансии DS-junior"                       
                from public.vacancies v
                where (lower(v.name) like '%data scientist%' or 
                lower(v.name) like '%data science%' or
                lower(v.name) like '%исследователь данных%' or
                (v.name like '%ML%' and v.name not like '%HTML%') or
                lower(v.name) like '%machine learning%' or
                lower(v.name) like '%машинн%обучен%%')
                and
               (v.experience like '%Нет опыта%' or
                v.experience like '%Стажировка%'or
                lower(v.name) like '%junior%')
                
             ''' 

In [67]:
# результат запроса
vac_es = pd.read_sql_query(query_6_2, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\1146784756.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_6_2, connection)


Вакансии DS-junior
0                  51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [122]:
# текст запроса
query_6_3 = f'''select
                    count(v.id) "Вакансии SQL или postgres"                       
                from public.vacancies v
                where (lower(v.name) like '%data scientist%' or 
                lower(v.name) like '%data science%' or
                lower(v.name) like '%исследователь данных%' or
                (v.name ilike '%ML%' and v.name not ilike '%HTML%') or -- поставил ilike т.к. не проходит задание
                lower(v.name) like '%machine learning%' or
                lower(v.name) like '%машинн%обучен%%')
                and 
                (v.key_skills like '%SQL%' or 
                v.key_skills like '%postgres%')
                
             ''' 

In [123]:
# результат запроса
vac_es = pd.read_sql_query(query_6_3, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\3688789454.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_6_3, connection)


Вакансии SQL или postgres
0                        229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [126]:
# текст запроса
query_6_4 = f'''select
                    count(v.id) "Вакансии Python"                       
                from public.vacancies v
                where (v.name ilike '%data scientist%' or 
                v.name ilike '%data science%' or
                v.name ilike '%исследователь данных%' or
                (v.name ilike '%ML%' and v.name not ilike '%HTML%') or -- поставил ilike т.к. не проходит задание 
                v.name ilike '%machine learning%' or
                v.name ilike '%машинн%обучен%%')       
                and 
                (v.key_skills ilike '%python%')
                
             '''  

In [127]:
# результат запроса
vac_es = pd.read_sql_query(query_6_4, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\3618162411.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_6_4, connection)


Вакансии Python
0              357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [94]:
# текст запроса
query_6_5 = f'''select
                    round(avg(length(v.key_skills) - length(replace(v.key_skills, chr(9), '')) + 1), 2) "Среднее к-во"                     
                from public.vacancies v
                where (lower(v.name) like '%data scientist%' or 
                lower(v.name) like '%data science%' or
                lower(v.name) like '%исследователь данных%' or
                (v.name like '%ML%' and v.name not like '%HTML%') or
                lower(v.name) like '%machine learning%' or
                lower(v.name) like '%машинн%обучен%%')  

            '''

In [114]:
# результат запроса
vac_es = pd.read_sql_query(query_6_5, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\2153957667.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_6_5, connection)


Среднее к-во
0          6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [104]:
# текст запроса
query_6_6 = f'''select
                     round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to))) "Средняя ЗП DS с опытом 3-6 лет"
                from public.vacancies v
                where (lower(v.name) like '%data scientist%' or 
                lower(v.name) like '%data science%' or
                lower(v.name) like '%исследователь данных%' or
                (v.name like '%ML%' and v.name not like '%HTML%') or
                lower(v.name) like '%machine learning%' or
                lower(v.name) like '%машинн%обучен%%') 
                and
                (v.salary_from is not null or salary_to is not null)   
                and 
                (v.experience = 'От 3 до 6 лет')    
                                
             '''   

In [107]:
# результат запроса
vac_es = pd.read_sql_query(query_6_6, connection)
vac_es

C:\Users\Володька\AppData\Local\Temp\ipykernel_22244\4179584256.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  vac_es = pd.read_sql_query(query_6_6, connection)


Средняя ЗП DS с опытом 3-6 лет
0                        243115.0

***

In [53]:
# выводы по предметному анализу

# выводы по предметному анализу
Вакансий, связанных с данными, 1771, при этом для junior DS из общего числа только 51 вакансия. 229 вакансий для DS требует в качестве ключевого навыка SQL или postgres, а python 357 вакансий. В вакансиях для DS в среднем указывают более 6 ключевых навыков. DS с опытом работы 3-6 лет может расчитывать в среднем на ЗП в 243115 руб.

# Общий вывод по проекту

In [54]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

# общий вывод по проекту
В выборке представлены 49197 вакансий от 23501 работодателей в 1362 регионах по 294 сферам деятельности.
Лидерами среди регионов по количеству размещенных вакансий являются регионы - мегаполисы с развитой промышленностью и развитой хозяйственной инфраструктурой.
Среднее значение по нижней границе ЗП - 71065, по верхней границе ЗП - 110537. Наиболее востребованные работники - работники на полный рабочий день с полной занятностью с опытом работы от 1 до 3 лет.
По количеству вакансий лидирует компания "Яндекс" - 1933 вакансии примерно 4% от общего количества. По количеству регионов в которых размещены вакансии компании, ожидаемо лидирует "Яндекс", на втором месте компания "Ростелеком". Возможно это связано с тем, что сферы деятельности "Яндекса" в основном связаны сIT и это позволяет иметь достаточно большой штат сотрудников на "удаленке" в различных регионах и даже странах, а у "Ростетелекома" развитая сеть представительств (которые физически существуют) по всей России.
У примерно 36% работодателей не указана сфера деятельности. Количество компаний у которых указано 4 и более сферы деятельности 2415; компаний, занимающихся разработкой ПО, 3553 - более 15%.
Вакансий, связанных с данными, 1771, при этом для junior DS из общего числа только 51 вакансия. 229 вакансия для DS требует в качестве ключевого навыка SQL или postgres, а python 357 вакансия. В вакансиях для DS в среднем указывают более 6 ключевых навыков. DS с опытом работы 3-6 лет может расчитывать в среднем на ЗП в 243115.